### PR TITLE
🚨🚨🚨  Fix rescale ViVit Efficientnet

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Fixes the rescaling logic for both EffiicentNet and ViVit. 

EfficientNet: The values were being rescale between [-0.5, 0.5] 

ViVit: values were being rescale between [-7.689350249903882e-06, 0.9999923106497501] as scale was being treated as having a value 255 in the `rescale` method, rather than 1/255. 

**This is a breaking change** and will affect the model outputs for both these models. However, it is a bug fix and should improve model predictions. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: tests/models/vivit/test_modeling_vivit.py [hunk 1]
```diff
@@ -345,6 +345,6 @@ def test_inference_for_video_classification(self):
         self.assertEqual(outputs.logits.shape, expected_shape)
 
         # taken from original model
-        expected_slice = torch.tensor([-1.0543, 2.0764, -0.2104, 0.4439, -0.9658]).to(torch_device)
+        expected_slice = torch.tensor([-0.9498,  2.7971, -1.4049,  0.1024, -1.8353]).to(torch_device)
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:348]: Integration test values didn't need to be updated for EfficientNet as the model used [doesn't offset the values when rescaling by default](https://github.com/huggingface/transformers/blob/dd9d45b6ecc0861847e21d461187711331a56138/src/transformers/models/efficientnet/image_processing_efficientnet.py#L92)

File: src/transformers/image_transforms.py [hunk 2]
```diff
@@ -110,10 +110,11 @@ def rescale(
     if not isinstance(image, np.ndarray):
         raise ValueError(f"Input image must be of type np.ndarray, got {type(image)}")
 
+    image = image.astype(np.float32)
```
Comment (amyeroberts) [inline src/transformers/image_transforms.py:113]: Image should be cast before scaling to ensure precision isn't lost when rescaling with a float value. 

This has the potential to slightly change the resulting pixel values for models. To check this doesn't affect our current models' outputs I ran integration tests for swin, vit, detr, clip and resnet.

File: src/transformers/image_transforms.py [hunk 3]
```diff
@@ -110,10 +110,11 @@ def rescale(
...
+
     rescaled_image = image * scale
     if data_format is not None:
         rescaled_image = to_channel_dimension_format(rescaled_image, data_format)
-    rescaled_image = rescaled_image.astype(dtype)
```
Comment (ydshieh) [inline src/transformers/image_transforms.py:116]: This makes the argument `dtype` unused. Is this intended?

**Review by ydshieh — APPROVED** [review]
Having a nit question about unused argument, otherwise LGTM.

Would like to have a review from @sgugger for a breaking change like this though.

File: src/transformers/image_transforms.py (hunk 2 above)
Comment (amyeroberts) [inline src/transformers/image_transforms.py:113]:
```suggestion
    image = image.astype(dtype)
```

File: src/transformers/image_transforms.py (hunk 3 above)
Comment (amyeroberts) [reply src/transformers/image_transforms.py:116]: Nope. Fixed now

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

File: tests/pipelines/test_pipelines_zero_shot_image_classification.py [hunk 4]
```diff
@@ -85,6 +85,7 @@ def test_small_model_pt(self):
             [
                 [{"score": 0.333, "label": "a"}, {"score": 0.333, "label": "b"}, {"score": 0.333, "label": "c"}],
                 [{"score": 0.333, "label": "a"}, {"score": 0.333, "label": "c"}, {"score": 0.333, "label": "b"}],
+                [{"score": 0.333, "label": "b"}, {"score": 0.333, "label": "a"}, {"score": 0.333, "label": "c"}],
```
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_zero_shot_image_classification.py:88]: @ydshieh I had to add another possible permutation here. I considered it OK as the comment above mentions the order isn't guaranteed.

Comment (ydshieh) [reply tests/pipelines/test_pipelines_zero_shot_image_classification.py:88]: No problem! Thanks for informing me.

**Comment by stas00** [comment]
If I'm not mistaken this PR is problematic as it breaks the `rescale` function. If I rebase my PR https://github.com/huggingface/transformers/pull/24796 to `main` the torchvision to transformers equivalency test fails: `tests/models/idefics/test_image_processing_idefics.py::IdeficsImageProcessingTest::test_torchvision_numpy_transforms_equivalency`

**Comment by rwightman** [comment]
@amyeroberts I'm guessing this was issue causing bad efficientnet predicons that I observed? 

FWIW I haven't ecountered any vision models that require separate codepaths for rescale vs mean/std [-1, 1] is mean 0.5, std 0.5, and [0, 1] is mean 0, std 1. ...

**Comment by amyeroberts** [comment]
@ydshieh I managed to track down the issue with EfficientNet down to the softmax which was being applied to the logits #25501 (sorry, I should have linked here as well). 

@rwightman Agreed, the design was something I either suggested or approved, so my bad. The good thing is, using `offset` isn't necessary, and users can simply pass in different means and stds to achieve the desired rescaling as you suggested.
